sources/imdsclient/src/lib.rs

```diff
@@ -0,0 +1,212 @@
+    session_token: String,
+}
+
+impl ImdsClient {
+    pub fn new() -> Result<Self> {
+        let client = Client::new();
+        let session_token = fetch_token(&client)?;
+        Ok(Self {
+            client,
+            session_token,
+        })
+    }
+
+    /// Helper to fetch `dynamic` targets from IMDS, preferring an override file if present.
+    pub fn fetch_dynamic(&mut self, end_target: &str, description: &str) -> Result<Option<String>> {
+        let dynamic_target = format!("dynamic/{}", end_target);
+        self.fetch_imds(IMDS_PINNED_DATE, &dynamic_target, description)
+    }
+
+    /// Helper to fetch `meta-data` targets from IMDS, preferring an override file if present.
+    pub fn fetch_metadata(
+        &mut self,
+        end_target: &str,
+        description: &str,
+    ) -> Result<Option<String>> {
+        let metadata_target = format!("meta-data/{}", end_target);
+        self.fetch_imds(IMDS_PINNED_DATE, &metadata_target, description)
+    }
+
+    /// Helper to fetch `user-data` from IMDS, preferring an override file if present.
+    pub fn fetch_userdata(&mut self) -> Result<Option<String>> {
+        self.fetch_imds(IMDS_PINNED_DATE, "user-data", "user-data")
+    }
+
+    /// Fetch data from IMDS, preferring an override file if present.
+    pub fn fetch_imds(
+        &mut self,
+        schema_version: &str,
+        target: &str,
+        description: &str,
+    ) -> Result<Option<String>> {
+        let uri = format!("{}/{}/{}", IMDS_BASE_URI, schema_version, target);
+        debug!("Requesting {} from {}", description, &uri);
+        let mut attempt: u16 = 0;
+        loop {
+            attempt += 1;
+            ensure!(attempt <= 2, error::FailedFetch { attempt });
+            let response = self
+                .client
+                .get(&uri)
+                .header("X-aws-ec2-metadata-token", &self.session_token)
+                .send()
+                .context(error::Request {
+                    method: "GET",
+                    uri: &uri,
+                })?;
+            trace!("IMDS response: {:?}", &response);
+
+            match response.status() {
+                code @ StatusCode::OK => {
+                    info!("Received {}", description);
+                    let response_body = response.text().context(error::ResponseBody {
+                        method: "GET",
+                        uri: &uri,
+                        code,
+                    })?;
+                    trace!("Response text: {:?}", &response_body);
+
+                    return Ok(Some(String::from(response_body.trim_end())));
+                }
+
+                // IMDS returns 404 if no user data is given, or if IMDS is disabled
+                StatusCode::NOT_FOUND => return Ok(None),
+
+                // IMDS returns 401 if the session token is expired or invalid
+                StatusCode::UNAUTHORIZED => {
+                    warn!("Session token is invalid or expired");
```

Comment:
Maybe this is actually `info!` since it's not actually a problem and we are going to do the refresh. Sometimes when you're debugging something hard, warnings can lead you astray.